include/mbgl/renderer/layer_tweaker.hpp

```diff
@@ -78,14 +86,15 @@ class LayerTweaker {
 protected:
     std::string id;
     Immutable<style::LayerProperties> evaluatedProperties;
-    bool propertiesUpdated = true;
 
 #if MLN_RENDER_BACKEND_METAL
     // For Metal, whether a property is provided through attribtues or uniforms is specified in
     // a uniform buffer rather than by a shader compiled with different preprocessor definitions.
     std::vector<std::string> propertiesAsUniforms;
```

Comment:
This could benefit from `StringIndexer`, and could become a `std::unordered_set<StringIdentity>` and that could help with performance in `LayerTweaker::hasPropertyAsUniform` which is called per frame to build the `permutationUBO`

Reply:
This PR doesn't add `propertiesAsUniforms`, so I'd suggest that should be in a separate one.  
That optimization should apply all the way through, including the return of `readDataDrivenPaintProperties` and the arguments of `getOrCreateShader`

Reply:
Correct. I created an issue for that. #1687